### PR TITLE
physic: non exported change: decimal type internal representation.

### PR DESCRIPTION
**Pull request by NeuralSpaz**
Change decimal type internal representation of digits from []byte to uint64. 
Moves validation of digits to atod from doti.
Update tests for new decimal literal.
Add new test cases to atod().
Update Decimal Benchmarks.

This is a ~2x faster atod() and ~5x faster dtoi(), and helps with #303 remaining units.

### CODE REVIEW

**Comment by NeuralSpaz** [comment]
gohci

File: conn/physic/units.go [hunk 1]
```diff
@@ -1389,10 +1364,11 @@ func dtoi(d decimal, scale int) (int64, error) {
 }
 
 // Converts a string to a decimal form. The return int is how many bytes of the
-// string are numeric. The string may contain +-0 prefixes and arbitrary
-// suffixes as trailing non number characters are ignored.
-// Significant digits are stored without leading or trailing zeros, rather an
-// exponent is used.
+// string are considered numeric. The string may contain +-0 prefixes and
+// arbitrary suffixes as trailing non number characters are ignored.
+// Significant digits are stored without leading or trailing zeros, rather a
```
Comment (maruel) [inline conn/physic/units.go]: I don't understand what you mean with the first part, as it's stored as an int64 now.

File: conn/physic/units.go [hunk 2]
```diff
@@ -1502,21 +1478,45 @@ func atod(s string) (decimal, int, error) {
 		}
 	}
 
-	if dp > start && dp < end {
-		// Concatenate with out decimal point.
-		d.digits = s[start:dp] + s[dp+1:end]
-	} else {
-		d.digits = s[start:end]
+	for i := start; i < end; i++ {
+
```
Comment (maruel) [inline conn/physic/units.go]: `c := s[i]` and use this instead of continuously indexing.

File: conn/physic/units.go [hunk 3]
```diff
@@ -1502,21 +1478,45 @@ func atod(s string) (decimal, int, error) {
...
+						err: errOverflowsInt64Negative,
+					}
+				}
+				return decimal{}, 0, &parseError{
+					msg: maxUint64Str,
+					err: errOverflowsInt64,
+				}
+			}
+			d.base = check
+		} else {
```
Comment (maruel) [inline conn/physic/units.go]: } else if c != '.' {

File: conn/physic/units.go [hunk 4]
```diff
@@ -1502,21 +1478,45 @@ func atod(s string) (decimal, int, error) {
...
-		d.digits = s[start:end]
+	for i := start; i < end; i++ {
+
+		// Check that is is a digit.
+		if s[i] >= '0' && s[i] <= '9' {
+			// '0' = 0x30 '1' = 0x31 ...etc.
+			digit := s[i] - '0'
+			// *10 is decimal shift left.
+			d.base *= 10
+			check := d.base + uint64(digit)
```
Comment (maruel) [inline conn/physic/units.go]:
```
// Convert ascii digit into number.
check := d.base + uint64(c - '0')
```

File: conn/physic/units.go [hunk 5]
```diff
@@ -1502,21 +1478,45 @@ func atod(s string) (decimal, int, error) {
...
 	if !isPoint {
 		d.exp = exp
 	} else {
-		ttl := dp - start
-		length := len(d.digits)
-		if ttl > 0 {
-			d.exp = ttl - length
-		} else {
-			d.exp = ttl - length + 1
+		if dp > start && dp < end {
```
Comment (maruel) [inline conn/physic/units.go]: I don't understand this condition. Can you add a comment?

File: conn/physic/units.go [hunk 6]
```diff
@@ -1502,21 +1478,45 @@ func atod(s string) (decimal, int, error) {
...
+			end--
+		}
+		d.exp = (dp - start) - (end - start) + 1
+		if (dp - start) > 0 {
```
Comment (maruel) [inline conn/physic/units.go]: `if dp > start {`
Please comment too.

**Comment by maruel** [comment]
gohci

File: conn/physic/units.go (hunk 1 above)
Comment (NeuralSpaz) [reply conn/physic/units.go]: I'm trying to say 1000 is stored as base:1, exp:3.
0.001 stored as base: 1, exp: -3, 
and 12.34 as base: 1234 exp: -2

Comment (NeuralSpaz) [reply conn/physic/units.go]: ... and that there is a limit to the number of significant digits (int64) eg you can't have 1234567890123456790.1234567890123456789.
But you can store very large or very small numbers 16* 10^-34 (0.00000000000000000000000000000000016) etc

Comment (maruel) [reply conn/physic/units.go]: ahh now I understand.

File: conn/physic/units.go [hunk 7]
```diff
@@ -1502,21 +1478,46 @@ func atod(s string) (decimal, int, error) {
...
-		} else {
-			d.exp = ttl - length + 1
+		if dp > start && dp < end {
+			// Decimal Point is in the middle of a number.
+			end--
+		}
+		// Find the exponet based on decimal point distance from left and the
+		// length of the number.
+		d.exp = (dp - start) - (end - start)
+		if (dp - start) <= 0 {
```
Comment (maruel) [inline conn/physic/units.go]: dp <= start

Comment (NeuralSpaz) [reply conn/physic/units.go]: oh good one.

**Pull request merged**